include/swift/SIL/SILWitnessTable.h

```diff
@@ -89,7 +103,9 @@ class SILWitnessTable : public llvm::ilist_node<SILWitnessTable>,
     Method,
     AssociatedType,
     AssociatedTypeProtocol,
-    BaseProtocol
+    BaseProtocol,
+    // SWIFT_ENABLE_TENSORFLOW
+    AutoDiffAssociatedFunction
```

Comment:
I was about to leave a comment here that it seems better to make this an ordinary method that's keyed by a `SILDeclRef` that describes the relationship to the original method, and then I discovered that Slava specifically told you not to do it that way.  Maybe we need to talk about this.

I'm not sure we care about the *size* of `SILDeclRef`, and the *complexity* hit seems low, especially compared to adding new kinds of witness table entries and so forth.

Reply:
I'll appreciate it if you and @slavapestov et al reach a consensus and let us know! I'll also try to schedule a discussion offline about this.

Reply:
While I was trying it the `SILDeclRef` way, I noticed some difficulties:
* I'm not sure if "autodiff associated function" is completely orthogonal to all the kinds and fields in `SILDeclRef`. e.g. what if we want separate "autodiff associated functions" for `Allocator` and `Initializer`? I haven't actually thought this through and found concrete problems -- just a worry that I have.
* `WitnessTableEntry`s in `ProtocolInfo.h` are keyed by the `Decl*`, so they break when there are multiple `SILDeclRef`s in the witness table with the same underlying `Decl*`. This was pretty easy to fix by adding more data to the `WitnessTableEntry` key. But maybe there are other places that I have not yet found that also assume that all their `SILDeclRef`s have distinct `Decl*`s.

I don't have a strong opinion about whether we should do it one way or the other. Both seem to work. Just letting you know about some things that I ran into while doing it with `SILDeclRef`.

Reply:
Oh also a third difficulty:
* A few of places assume that the `SILDeclRef` refers to something with the same type as the underlying `Decl*`. Adding "autodiff associated functions" breaks this assumption. It was pretty easy to adjust all the places that I found that mattered for this PR, but I'm not sure if there are other places lurking that are harder to find or use the assumption in more subtle ways.

Reply:
Do you remember an example of such a place?  That would be very surprising since e.g. foreign and non-foreign `SILDeclRef`s obviously have different types.

Reply:
> * I'm not sure if "autodiff associated function" is completely orthogonal to all the kinds and fields in `SILDeclRef`. e.g. what if we want separate "autodiff associated functions" for `Allocator` and `Initializer`? I haven't actually thought this through and found concrete problems -- just a worry that I have.

I don't think it should be orthogonal.  I wouldn't add a new kind at all: you should add the ability to add a derivation path to an arbitrary other `SILDeclRef`.

> * `WitnessTableEntry`s in `ProtocolInfo.h` are keyed by the `Decl*`, so they break when there are multiple `SILDeclRef`s in the witness table with the same underlying `Decl*`. This was pretty easy to fix by adding more data to the `WitnessTableEntry` key. But maybe there are other places that I have not yet found that also assume that all their `SILDeclRef`s have distinct `Decl*`s.

Hmm, that's unfortunate.  I don't know why they would need to break down the `SILDeclRef` like that.

Anyway, Slava and I will talk about it.

Reply:
The `WitnessTableEntry` thing is probably just divergent evolution we never patched up when SIL was added in. It should almost definitely just use `SILDeclRef` now.

Reply:
> Do you remember an example of such a place? That would be very surprising since e.g. foreign and non-foreign SILDeclRefs obviously have different types.

Some places are:
* Search `getDecl()->getInterfaceType()` in [SILPrinter.cpp](https://github.com/apple/swift/blob/01bfdc162e2bfdaa609e25a30c3b6b965fccd272/lib/SIL/SILPrinter.cpp).
* [This lookup logic in ParseSIL.cpp](https://github.com/apple/swift/blob/master/lib/ParseSIL/ParseSIL.cpp#L1928).

Actually, I think that is all that I found. Maybe it does make sense to do it that way in SILPrinter/ParseSIL if you think of the type in the printed SIL as always being the type of the underlying decl.

Reply:
No, those are both correct.  The type there is not the type of the constant identified by the `SILDeclRef` — it's the type of the declaration that serves as the root of the `SILDeclRef`, and it's only there to potentially resolve between overloads.

Reply:
That makes sense. So the third difficulty I listed is not actually a problem.

Reply:
Slava and I talked about this after the meeting today and decided that the `SILDeclRef` approach makes sense.  The two kinds of derivative should each be `SILDeclRef::Kind`s, and you should be able to re-use one of the other fields (like `defaultArgIndex`) for the order of the derivative.

...it's quite unfortunate that "higher-order" is used for both repeated derivation and functions of functions.

Reply:
Awesome. Thanks John!